include/tsar/Analysis/Memory/MemorySetInfo.h

```diff
@@ -157,6 +192,93 @@ template<> struct MemorySetInfo<MemoryLocationRange> {
   static inline MemoryLocationRange make(const MemoryLocationRange &Loc) {
     return MemoryLocationRange(Loc);
   }
+  static inline uint64_t getNumDims(
+      const MemoryLocationRange &Loc) noexcept {
+    return Loc.DimList.size();
+  }
+  static inline bool areJoinable(
+      const MemoryLocationRange &LHS, const MemoryLocationRange &RHS) {
+    if (LHS.DimList.size() != RHS.DimList.size())
+      return false;
+    if (LHS.DimList.size() == 0) {

```

Comment:
Для вектора это не так важно, но в целом `empty()` часто эффективнее, чем вычисление размера.

```suggestion
    if (LHS.DimSize.empty()) {

```

Reply:
Отвечу здесь не на этот ваш комментарий, а на тот, что чуть выше.

> Посмотрел еще часть, не смотрел пока непосредственно реализацию `intersect`.
> 
> Такой вопрос, сейчас когда хранятся участки памяти в `MemorySet`, обобщенные и не обобщенные хранятся фактически независимо? Т,е. операции между обобщенным и не обобщенным участком фактически никакие не возможны? Я не предлагаю это сейчас делать, но в теории их можно было бы хранить отдельно (список обобщенных и список не обобщенных участков для одного `Ptr`)? Или так не получится?

Сейчас все устроено так, что `Ptr` у обобщенного с необобщенного участка памяти будет различным. При этом для обобщенных участков, которые относятся к одному и тому же массиву, `Ptr = ArrayPtr->getBase()` (`ArrayPtr` берется из результатов делинеаризации). Общий указатель для обобщенных участков нужен для того, чтобы между ними можно было проводить операции. Иначе не будет понятно, к одному массиву относятся участки или нет, потому что все необобщенные `a[i]` имеют разный `Ptr`. Мне пока не приходит в голову, как можно привязывать обобщенные участки к необобщенным так, чтобы сохранить возможность проводить операции между необобщенными.

А имеет ли какой-то смысл проведение операции между обобщенным и необобщенным `Loc`? В текущей реализации обобщенные участки памяти выступают лишь как вспомогательные, по ним выбираются ExplicitAccess'ы для добавления в множества.

Reply:
Я как раз скорее склоняюсь к тому, что диапазоны и поэлементные участки памяти - независимые сущности, и может быть их вообще надо было бы хранить по отдельности. Но это пока на стадии размышлений, о том, не окажет ли влияние изменение `MemoryLocationRange` на какие-то другие проходы. Пока что оставляем как есть. Сначала, главное чтобы заработало то, что уже реализовано.